### ISSUE TITLE
function generic type doesn't throw error when has extra property

### ISSUE THREAD

**Issue by troy351** [microsoft/TypeScript#48219]
# Bug Report

### 🕗 Version & Regression Information

v4.6.2

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=4.6.2#code/C4TwDgpgBAslC8UDeUDaBrCIBcUDOwATgJYB2A5gLq4oA2AhgEYS24EkVQC+3AsAFACAJhADGDQtABmAV1KjgxAPakoogBZjMhADwAVKBAAewCKSF5YAPgAUo3HoCUuAG5LiQgQI1aIhG0gCUMFQ9DRQDMysUADk9DEANFBSSkq4RDLQXEkA9Dn46koytEJQwOqESgDuhoSVhFCMMsBQpErASeV+0MSW9KrGRPRQAAYpSiNQYJWQhKBBIYzhkSy4MYwx3AkCXI5e-D6i2gELwWHIEUyrsfFJ4+mEmVtQeVCaklBCHmUV1ZZ+9VC-yMkAUECESSaLSET2ASh+0FIg1GjEmDCOliUUlGK1oI1OjRo2R2e0EZIEZFMhCk9FE0AAgsgCRgsLgRgBadkAEiQ7DI5C4I2o+CI-J2+1EKgIoVwjMQgX4IVinPia1uBPWMTWGxe+XKvR+lSq-zqSga9GBoNMnn4XH2InE9A+snkihUak0Rz8ACZ9IYTGYLFB6bYXPRok5XO4bd5PdpvSdFSEYiqtTdEhqNtrNq99ZYhEoIJYAHIAeQM5SNJsBFv9VvBJKAA

### 💻 Code

```ts
type M = { [key: string]: { label: string } }

declare function checker<T extends M>(c: T): void

checker({
    a: { label: 'a', foo: true }, // should throw error but not, there is an extra `foo` property
    b: { label: 'b' },
})

checker({
    a: { label: 'a', foo: true }, // here did throws error as expected, but due to the next `b` lacks of `label`
    b: {},
})



interface A {
    [key: `--${string}`]: string
}

const a: A = {
    '--a': 'a',
    'b': 'b' // this throws error as expected
}

declare function checker2<T extends A>(val: T): void

checker2({
    '--a': 'a',
    'b': 'b' // this does NOT throws error as expected
})
```

**Comment by MartinJohns** [comment]
![image](https://user-images.githubusercontent.com/5269069/157860546-c3b2c820-9279-41cb-8600-e33ddcaf3fbf.png)
Well, what is it?

You want #12936.

**Comment by RyanCavanaugh** [comment]
The entire point of generics is to express behavior over subtypes; wanting to disallow subtypes implies you don't want the function to be generic in the first place.

**Comment by troy351** [comment]
But I do want to using generic to infer some type, like `keyof` of the argument object while keeping the key of argument object matches template literal

**Comment by RyanCavanaugh** [comment]
This is already possible:
```ts
declare function checker2<T extends `--${string}`>(val: Record<T, string>): void

checker2({
    '--a': 'a',
    'b': 'b' // Errors
})
```

**Comment by RyanCavanaugh** [comment]
I don't quite understand what you're looking for with that; more examples might help but you will likely get more detailed answers on a help-oriented forum (discord, Stack Overflow)

**Comment by troy351** [comment]
I figured it out
```ts
declare function checker2<T extends `--${string}`>(val: { cssVars: Record<T, string>, cssVarOverwrite: Partial<Record<T, string>> }): void
```

Thanks for your example
